File: System_H/SSDR.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Xml;
using HSFUniverse;
using MissionElements;
using Utilities;
using HSFSystem;

namespace HSFSubsystem
{
    public class SSDR : Subsystem
    {
        //Some Defaults
        private double _bufferSize = 4098;
        public static StateVarKey<double> DATABUFFERRATIO_KEY; 

        public SSDR(XmlNode SSDRXmlNode, Dependencies dependencies, Asset asset)
        {
            DefaultSubName = "SSDR";
            Asset = asset;
            getSubNameFromXmlNode(SSDRXmlNode);
            if (SSDRXmlNode.Attributes["bufferSize"] != null)
                _bufferSize = (double)Convert.ChangeType(SSDRXmlNode.Attributes["bufferSize"].Value.ToString(), typeof(double));
            DATABUFFERRATIO_KEY = new StateVarKey<double>(Asset.Name + "." +"DataBufferFillRatio");
            addKey(DATABUFFERRATIO_KEY);
            SubsystemDependencyFunctions = new Dictionary<string, Delegate>();
            DependentSubsystems = new List<ISubsystem>();
            dependencies.Add("PowerfromSSDR", new Func<Event, HSFProfile<double>>(POWERSUB_PowerProfile_SSDRSUB));
            dependencies.Add("CommfromSSDR", new Func<Event, HSFProfile<double>>(COMMSUB_DataRateProfile_SSDRSUB));
        }

        /// <summary>
        /// An override of the subsystem canPerform method
        /// </summary>
        /// <param name="oldState"></param>
        /// <param name="newState"></param>
        /// <param name="tasks"></param>
        /// <param name="environment"></param>
        /// <returns></returns>
        public override bool canPerform(Event proposedEvent, Universe environment)
        {
            if (!base.canPerform(proposedEvent, environment))
                return false;
            if (_task.Type == TaskType.IMAGING)
            {
                double ts = proposedEvent.GetTaskStart(Asset);
                double te = proposedEvent.GetTaskEnd(Asset);
                double oldbufferratio = _oldState.getLastValue(Dkeys.First()).Value;
                HSFProfile<double> newdataratein = (DependencyCollector(proposedEvent) / _bufferSize);

                bool exceeded = false;
                HSFProfile<double> newdataratio = newdataratein.upperLimitIntegrateToProf(ts, te, 5, 1, ref exceeded, 0, oldbufferratio);
                if (!exceeded)
                {
                    _newState.addValue(DATABUFFERRATIO_KEY, newdataratio);
                    return true;
                }
                Logger.Report("SSDR");
                return false;
            }
            else if (_task.Type == TaskType.COMM)
            {
                double ts = proposedEvent.GetTaskStart(Asset);
                proposedEvent.SetTaskEnd(Asset, ts + 60.0);
                double te = proposedEvent.GetTaskEnd(Asset);

                double data = _bufferSize * _oldState.getLastValue(Dkeys.First()).Value;
                double dataqueout = data / 2 > 50 ? data / 2 : data;

                if (data - dataqueout < 0)
                    dataqueout = data;

                if (dataqueout > 0)
                    _newState.addValue(DATABUFFERRATIO_KEY, new KeyValuePair<double, double>(te, (data - dataqueout) / _bufferSize));
                return true;
            }
            return true;
        }

        /// <summary>
        /// Dependecy Function for the power subsystem
        /// </summary>
        /// <param name="currentState"></param>
        /// <returns></returns>
        HSFProfile<double> POWERSUB_PowerProfile_SSDRSUB(Event currentEvent)
        {
            HSFProfile<double> prof1 = new HSFProfile<double>();
            prof1[currentEvent.GetEventStart(Asset)] = 15;
            return prof1;
        }

        /// <summary>
        /// Dependecy function for the ssdr subsystem
        /// </summary>
        /// <param name="currentEvent"></param>
        /// <returns></returns>
        HSFProfile<double> COMMSUB_DataRateProfile_SSDRSUB(Event currentEvent)
        {
            double datarate = 5000 * (currentEvent.State.getValueAtTime(DATABUFFERRATIO_KEY, currentEvent.GetTaskStart(Asset)).Value - currentEvent.State.getValueAtTime(DATABUFFERRATIO_KEY, currentEvent.GetTaskEnd(Asset)).Value) / (currentEvent.GetTaskEnd(Asset) - currentEvent.GetTaskStart(Asset));
            HSFProfile<double> prof1 = new HSFProfile<double>();
            if (datarate != 0)
            {
                prof1[currentEvent.GetTaskStart(Asset)] = datarate;
                prof1[currentEvent.GetTaskEnd(Asset)] = 0;
            }
            return prof1;
        }
    }
}

File: ssdr.py
import sys
import clr
import System.Collections.Generic
import System
clr.AddReference('System.Core')
clr.AddReference('IronPython')
clr.AddReference('System.Xml')
clr.AddReferenceByName('Utilities')
clr.AddReferenceByName('HSFUniverse')
clr.AddReferenceByName('UserModel')
clr.AddReferenceByName('MissionElements')
clr.AddReferenceByName('HSFSystem')

import System.Xml
import HSFSystem
import HSFSubsystem
import MissionElements
import Utilities
import HSFUniverse
import UserModel
from HSFSystem import *
from System.Xml import XmlNode
from Utilities import *
from HSFUniverse import *
from UserModel import *
from MissionElements import *
from System import Func, Delegate
from System.Collections.Generic import *
from IronPython.Compiler import CallTarget0

class ssdr(HSFSubsystem.Subsystem):
    def __new__(cls, node, asset):
        instance = HSFSubsystem.Subsystem.__new__(cls)
        instance.Asset = asset
        instance.Name = instance.Asset.Name + '.' + node.Attributes['subsystemName'].Value.ToString().ToLower()

        if (node.Attributes['bufferSize'] != None):
            instance._bufferSize = float(node.Attributes['bufferSize'].Value.ToString())
        instance.DATABUFFERRATIO_KEY = Utilities.StateVariableKey[System.Double](instance.Asset.Name + '.' + 'databufferfillratio')
        instance.addKey(instance.DATABUFFERRATIO_KEY)
        return instance

    def CanPerform(self, event, universe):
        #print(self._task.Type)
        if (self._task.Type == "imaging"):
            ts = event.GetTaskStart(self.Asset)
            te = event.GetTaskEnd(self.Asset)
            oldbufferratio = self._newState.GetLastValue(self.Dkeys[0]).Item2
            newdataratein = HSFProfile[System.Double]()
            newdataratein = self.DependencyCollector(event) / self._bufferSize
            exceeded = False
            newdataratio = HSFProfile[System.Double]()
            newdataratio = newdataratein.upperLimitIntegrateToProf(ts, te, 1, 1, exceeded, 0, oldbufferratio)
            if (exceeded == False):
                self._newState.AddValues(self.DATABUFFERRATIO_KEY, newdataratio[0])
                #print("SSDR CanPreform, Imaging Task, Dataratio")
                #print(newdataratio[0])
                return True
            return False
        if(self._task.Type == "comm"):
            print("Entry of SSDR Comm Task")
            ts = event.GetTaskStart(self.Asset)
            event.SetTaskEnd(self.Asset, ts + 60.0)
            te = event.GetTaskEnd(self.Asset)
            data = self._bufferSize * self._newState.GetLastValue(self.Dkeys[0]).Item2
            if( data / 2 > 50):
                dataqueout = data/2
            else:
                dataqueout = data
            if (data - dataqueout < 0):
                dataqueout = data
            if (dataqueout > 0):
                self._newState.AddValue(self.DATABUFFERRATIO_KEY, te, (data - dataqueout) / self._bufferSize)
            print("SSDR CanPreform, Comm Task")
            return True
        return True

    def CanExtend(self, event, universe, extendTo):
        return super(ssdr, self).CanExtend(event, universe, extendTo)

    def Power_asset1_from_SSDR_asset1(self, event):
        prof1 = HSFProfile[System.Double]()
        prof1[event.GetEventStart(self.Asset)] = 15
        return prof1

    def Comm_asset1_from_SSDR_asset1(self, event):
        datarate = 5000 * (event.State.GetValueAtTime(self.DATABUFFERRATIO_KEY, event.GetTaskStart(self.Asset)).Item2 - event.State.GetValueAtTime(self.DATABUFFERRATIO_KEY, event.GetTaskEnd(self.Asset)).Item2) / (event.GetTaskEnd(self.Asset) - event.GetTaskStart(self.Asset))
        prof1 = HSFProfile[System.Double]()
        if (datarate != 0):
            prof1[event.GetTaskStart(self.Asset)] = datarate
            prof1[event.GetTaskEnd(self.Asset)] = 0
        return prof1

    def DepFinder(self, depFnName):  # Search for method from string input
        fnc = getattr(self, depFnName)
        dep = Dictionary[str, Delegate]()
        depFnToAdd = Func[Event, Utilities.HSFProfile[System.Double]](fnc)
        dep.Add(depFnName, depFnToAdd)
        return dep

    def GetDependencyCollector(self):
        return Func[Event,  Utilities.HSFProfile[System.Double]](self.DependencyCollector)

    def DependencyCollector(self, currentEvent):
        return super(ssdr, self).DependencyCollector(currentEvent)

#    def GetDependencyDictionary(self):
#        dep = Dictionary[str, Delegate]()
#        depFunc1 = Func[Event,  Utilities.HSFProfile[System.Double]](self.POWERSUB_PowerProfile_SSDRSUB)
#        depFunc2 = Func[Event,  Utilities.HSFProfile[System.Double]](self.COMMSUB_DataRateProfile_SSDRSUB)
#        dep.Add("CommfromSSDR" + "." + self.Asset.Name, depFunc2)
#        depFunc3 = Func[Event,  System.Double](self.EVAL_DataRateProfile_SSDRSUB)
#        dep.Add("EvalfromSSDR" + "." + self.Asset.Name, depFunc3)
#        return dep

#    def EVAL_DataRateProfile_SSDRSUB(self, event):
#        return (event.State.GetValueAtTime(DATABUFFERRATIO_KEY, event.GetTaskEnd(self.Asset)).Value - event.State.GetValueAtTime(DATABUFFERRATIO_KEY, event.GetTaskEnd(self.Asset)).Value) * 50
